compiler/daml-lf-proto/src/DA/Daml/LF/Proto3/EncodeV1.hs

```diff
@@ -562,19 +562,7 @@ encodeDefName version txt = case mangleIdentifier txt of
 
 -- | NOTE(MH): This functions is used for sanity checking. The actual checks
 -- are done in the conversion to DAML-LF.
-checkFeature :: (Version -> Bool) -> String -> Version -> a -> a
-checkFeature hasFeature name version x
-    | hasFeature version = x
-    | otherwise = error (error "DAML-LF " ++ renderPretty version ++ " cannot encode feature: " ++ name)
-
-checkOptional :: Version -> a -> a
-checkOptional = checkFeature supportsOptional "Optional"
-
-checkArrowType :: Version -> a -> a
-checkArrowType = checkFeature supportsArrowType "Partial application of (->)"
-
-checkContractKeys :: Version -> a -> a
-checkContractKeys = checkFeature supportsContractKeys "Contract keys"
-
-checkTextMap :: Version -> a -> a
-checkTextMap = checkFeature supportsTextMap "TextMap"
+checkFeature :: Feature -> Version -> a -> a
+checkFeature feature version x
+    | version `supports` feature = x
+    | otherwise = error (error "DAML-LF " ++ renderPretty version ++ " cannot encode feature: " ++ T.unpack (featureName feature))
```

Comment:
Looks like this did not change in this PR but why are we walling error twice here?

Reply:
👍  - let's error at most once